event_sourcery/aggregate.py

```diff
@@ -1,27 +1,37 @@
-from typing import Any, Type
+from contextlib import contextmanager
+from typing import Any, Iterator, Type
 
 from event_sourcery.interfaces.event import Event
-from event_sourcery.types.stream_id import StreamId
 
 
 class Aggregate:
-    def __restore_aggregate_state__(
-        self, past_events: list[Event], changes: list[Event], stream_id: StreamId
-    ) -> None:
-        for event in past_events:
-            self._apply(event)
+    def __init__(self) -> None:
+        self.__version = 0
+        self.__changes: list[Event] = []
 
-        self.__changes = changes
-        self.__stream_id = stream_id
+    def __rehydrate__(self, event: Event) -> None:
+        self._apply(event)
+        self.__version = event.version
+
+    @contextmanager
+    def __persisting_changes__(self) -> Iterator[Iterator[Event]]:
```

Comment:
A context-manager used during saving the aggregate to guarantee that version would be upgraded and pending events cleared. So we can clearly save aggregate, modify it again and save again